test/src/history/code_history_controller_test.dart

```diff
@@ -252,73 +378,120 @@ void main() {
+        controller.value = controller.value.typed('b');
+        await wt.moveCursor(-1);
+        expect(controller.historyController.stack.length, 5);
+
+        await wt.sendUndo();
+        expect(controller.value.text, visible + 'bb');
+        //                                        \ selection
         expect(
-          controller.selection,
-          const TextSelection.collapsed(
-            offset: MethodSnippet.visible.length + 1,
-          ),
+          controller.value.selection,
+          TextSelection.collapsed(offset: controller.value.text.length - 1),
+        );
+        await wt.sendUndo();
+        expect(controller.value.text, visible + 'b');
+        //                                       \ selection
+        expect(
+          controller.value.selection,
+          TextSelection.collapsed(offset: controller.value.text.length - 1),
         );
-
         await wt.sendUndo();
+        expect(controller.value.text, visible + 'b');
+        //                                         \ selection
+        expect(
+          controller.value.selection,
+          TextSelection.collapsed(offset: controller.value.text.length),
+        );
         await wt.sendUndo();
+        expect(controller.value.text, visible);
+        //                                   \ selection
+        expect(
+          controller.value.selection.start,
+          controller.value.text.length,
+        );
 
-        expect(controller.fullText, MethodSnippet.full);
+        expect(controller.code.foldedBlocks.length, 1);
 
+        await wt.sendRedo();
+        await wt.sendRedo();
         await wt.sendRedo();
         await wt.sendRedo();
 
-        expect(controller.fullText, MethodSnippet.full + 'ab');
+        expect(controller.code.foldedBlocks.length, 1);
       });
 
-      testWidgets('Selection does not disable redo', (WidgetTester wt) async {
+      testWidgets('Selection disables redo', (WidgetTester wt) async {
         final controller = await pumpController(wt, MethodSnippet.full);
 
         await wt.cursorEnd();
         controller.value = controller.value.typed('a');
         await wt.sendUndo(); // Creates.
+        expect(controller.historyController.stack.length, 2);
         await wt.pumpAndSettle();
 
         controller.value = controller.value.copyWith(
           selection: const TextSelection.collapsed(offset: 0),
         );
-        expect(controller.historyController.stack.length, 2);
+        expect(controller.historyController.stack.length, 1);
       });
 
       testWidgets('Limit depth', (WidgetTester wt) async {
         final controller = await pumpController(wt, MethodSnippet.full);
         await wt.cursorEnd();
 
         // 1. Fill the limit.
-
-        for (int i = 0; i < CodeHistoryController.limit - 1; i++) {
+        for (int i = 0; i < CodeHistoryController.limit / 2 - 1; i++) {
           controller.value = controller.value.typed('a');
           await wt.moveCursor(-1); // Creates.
         }
```

Comment:
Insert `ab` on each iteration so that it creates exactly 1 record. This way we avoid dependency on `limit` being even.